File: apps/frontend/server/routes/discovery.ts
import { Router, Request, Response } from 'express'
import { proxyToBackend } from '../../lib/proxy'

const router = Router()

// Convert a snake_case DiscoveredResource from the Go backend to camelCase
function mapResource(r: Record<string, unknown>) {
  return {
    id: r.id,
    ip: r.ip,
    port: r.port,
    protocol: r.protocol,
    serviceName: r.service_name || 'Unknown',
    reachableFrom: r.reachable_from,
    firstSeen: r.first_seen,
  }
}

// Convert a snake_case ScanJob from the Go backend to camelCase
function mapScanJob(job: Record<string, unknown>) {
  const results = Array.isArray(job.results)
    ? job.results.map((r: Record<string, unknown>) => mapResource(r))
    : undefined
  return {
    requestId: job.request_id,
    connectorId: job.connector_id,
    status: job.status,
    targets: job.targets,
    ports: job.ports,
    startedAt: job.started_at,
    completedAt: job.completed_at,
    results,
    error: job.error,
  }
}

// POST /api/discovery/scan — start a network discovery scan
router.post('/scan', async (req: Request, res: Response) => {
  try {
    const result = await proxyToBackend('/api/admin/discovery/scan', {
      method: 'POST',
      body: JSON.stringify(req.body),
    })
    res.json(result)
  } catch (error) {
    res.status(500).json({ error: (error as Error).message })
  }
})

// GET /api/discovery/scan/:requestId — get scan status
router.get('/scan/:requestId', async (req: Request, res: Response) => {
  try {
    const { requestId } = req.params
    const raw = await proxyToBackend(`/api/admin/discovery/scan/${requestId}`) as Record<string, unknown>
    res.json(mapScanJob(raw))
  } catch (error) {
    res.status(500).json({ error: (error as Error).message })
  }
})

// GET /api/discovery/results — get all discovered resources
router.get('/results', async (req: Request, res: Response) => {
  try {
    const connectorId = req.query.connector_id
    const query = connectorId ? `?connector_id=${connectorId}` : ''
    const raw = await proxyToBackend(`/api/admin/discovery/results${query}`) as Record<string, unknown>[]
    res.json(Array.isArray(raw) ? raw.map(mapResource) : [])
  } catch (error) {
    res.status(500).json({ error: (error as Error).message })
  }
})

export default router

File: services/controller/state/discovery.go
package state

import (
	"sync"
	"time"
)

// ScanStatus represents the current state of a scan job.
type ScanStatus string

const (
	ScanStatusPending    ScanStatus = "pending"
	ScanStatusInProgress ScanStatus = "in_progress"
	ScanStatusCompleted  ScanStatus = "completed"
	ScanStatusFailed     ScanStatus = "failed"
)

// DiscoveredResource represents a network resource found during scanning.
type DiscoveredResource struct {
	ID           string `json:"id"`
	IP           string `json:"ip"`
	Port         uint16 `json:"port"`
	Protocol     string `json:"protocol"`
	ServiceName  string `json:"service_name"`
	ReachableFrom string `json:"reachable_from"`
	FirstSeen    int64  `json:"first_seen"`
}

// ScanJob tracks a single discovery scan request.
type ScanJob struct {
	RequestID   string               `json:"request_id"`
	ConnectorID string               `json:"connector_id"`
	Status      ScanStatus           `json:"status"`
	Targets     []string             `json:"targets"`
	Ports       []uint16             `json:"ports"`
	StartedAt   time.Time            `json:"started_at"`
	CompletedAt *time.Time           `json:"completed_at,omitempty"`
	Results     []DiscoveredResource `json:"results,omitempty"`
	Error       string               `json:"error,omitempty"`
}

// ScanStore is a thread-safe in-memory store for scan jobs.
type ScanStore struct {
	mu   sync.RWMutex
	jobs map[string]*ScanJob
}

// NewScanStore creates a new empty ScanStore.
func NewScanStore() *ScanStore {
	return &ScanStore{
		jobs: make(map[string]*ScanJob),
	}
}

// Create adds a new scan job in pending state.
func (s *ScanStore) Create(job *ScanJob) {
	s.mu.Lock()
	defer s.mu.Unlock()
	job.Status = ScanStatusPending
	job.StartedAt = time.Now().UTC()
	s.jobs[job.RequestID] = job
}

// Get returns a scan job by request ID.
func (s *ScanStore) Get(requestID string) (*ScanJob, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	job, ok := s.jobs[requestID]
	return job, ok
}

// MarkInProgress marks a scan job as in-progress.
func (s *ScanStore) MarkInProgress(requestID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if job, ok := s.jobs[requestID]; ok {
		job.Status = ScanStatusInProgress
	}
}

// Complete marks a scan job as completed with results.
func (s *ScanStore) Complete(requestID string, results []DiscoveredResource) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if job, ok := s.jobs[requestID]; ok {
		now := time.Now().UTC()
		job.Status = ScanStatusCompleted
		job.CompletedAt = &now
		job.Results = results
	}
}

// Fail marks a scan job as failed with an error message.
func (s *ScanStore) Fail(requestID string, errMsg string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if job, ok := s.jobs[requestID]; ok {
		now := time.Now().UTC()
		job.Status = ScanStatusFailed
		job.CompletedAt = &now
		job.Error = errMsg
	}
}

// List returns all scan jobs.
func (s *ScanStore) List() []*ScanJob {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make([]*ScanJob, 0, len(s.jobs))
	for _, j := range s.jobs {
		result = append(result, j)
	}
	return result
}
